6690: New branch feature/6690_send_trigger_ddl

philipmarzullo64 opened the pull request:
this feature has been tested with the following platforms:
Oracle, SQL Server, MariaDB, MySQL, and Postgres.

Commit:
6690: New branch feature/6690_send_trigger_ddl

pavel-jm commented on symmetric-db/src/main/java/org/jumpmind/db/model/Function.java:
```diff
@@ -0,0 +1,171 @@
+        this.triggerName = triggerName;
+    }
+    
+    public void removePlatformFunction(String databaseName) {
```
This is confusing me: is the databaseName parameter actually referring to a name of a PlatformFunction?

philipmarzullo64 replied:
Yes, all of the Platform* objects have a name that represents the database platform. This is because the Platform* object is embedded in a non-Platform object, giving it context for what it represents. For example, PlatformFunction has a name of postgres, and this PlatformFunction is embedded in the Function object. The Function object has the name of the database object, and the embedded PlatformFunction object inherits the name of the Function. The name of the PlatformFunction is the database platform, allowing multiple PlatformFunction objects to be embedded in the single Function object.

pavel-jm commented on symmetric-db/src/main/java/org/jumpmind/db/model/Function.java:
```diff
@@ -0,0 +1,171 @@
+        return platformFunctions;
+    }
+
+    public PlatformFunction findPlatformFunction(String name) {
```
This is confusing me: does the name parameter refer to a name of a PlatformFunction or a database engine type?

philipmarzullo64 replied:
See comment on previous question.

pavel-jm commented on symmetric-db/src/main/java/org/jumpmind/db/model/Function.java:
```diff
@@ -0,0 +1,171 @@
+    private PlatformFunction findDifferentVersionPlatformFunction(String name) {
+        if (platformFunctions != null) {
+            for (Entry<String, PlatformFunction> entry : platformFunctions.entrySet()) {
+                if (entry.getKey() != null && entry.getKey().contains(name)) {
```
Is the .contains() filter selective enough? A customer might have two tables: "foo" and  "foo_2". 
Would this function search return "sym_i_trg_foo_2" when asked for  "sym_i_trg_foo"?

philipmarzullo64 replied:
This is looking for database platform types that have "versions" in them. For example, there are "postgres" and "postgres95" database platform types. This function is looking for different versions of the same platform type.

pavel-jm commented on symmetric-db/src/main/java/org/jumpmind/db/model/Function.java:
```diff
@@ -0,0 +1,171 @@
+    private String schemaName;
+    private String tableName;
+    private String triggerName;
+    Map<String, PlatformFunction> platformFunctions;
```
I crave a comment explaining why platformFunctions is here? What is the function of it, no pun intended.
 Or maybe a comment at the class level, explaining which function objects this is intended for?

philipmarzullo64 replied:
See reply on first comment, regarding the PlatformFunction objects being embedded inside a Function object. With that understanding, is a comment still needed?

pavel-jm replied:
Not a high priority, but new-ish developers will definitely appreciate it.

erilong commented on symmetric-db/src/main/java/org/jumpmind/db/alter/RemoveTriggerChange.java:
```diff
@@ -0,0 +1,26 @@
+
+public class RemoveTriggerChange extends TableChangeImplBase {
+    private Trigger _trigger;
+    //private 
```
Remove commented out code.

philipmarzullo64 replied:
Done

erilong commented on symmetric-db/src/main/java/org/jumpmind/db/model/Function.java:
```diff
@@ -0,0 +1,171 @@
+    }
+
+    public Function() {
+    };
```
Remove extraneous semi-colon.

philipmarzullo64 replied:
Done

erilong commented on symmetric-db/src/main/java/org/jumpmind/db/model/Function.java:
```diff
@@ -0,0 +1,171 @@
+    
+    public static String getFullyQualifiedName(String catalog, String schema, String tableName, String functionName) {
+        String fullName = "";
+        if (catalog != null)
```
Needs curly braces.

philipmarzullo64 replied:
Done

erilong commented on symmetric-db/src/main/java/org/jumpmind/db/model/Table.java:
```diff
@@ -1590,6 +1601,72 @@ public void setMadeAllColumnsPrimaryKey(boolean madeAllColumnsPrimaryKey) {
+        return triggers.toArray(new Trigger[triggers.size()]);
+    }
+
+    public List<Trigger> getCTriggersAsList() {
```
This looks like a typo.  Should be getTriggersAsList().

philipmarzullo64 replied:
Done

erilong commented on symmetric-db/src/main/java/org/jumpmind/db/platform/AbstractDatabasePlatform.java:
```diff
@@ -264,15 +274,33 @@ public void alterTables(boolean continueOnError, IAlterDatabaseInterceptor[] int
         String alterSql = ddlBuilder.alterDatabase(currentDatabase, desiredDatabase, interceptors);
         if (StringUtils.isNotBlank(alterSql.trim())) {
-            log.info("Running alter sql:\n{}", alterSql);
+            // log.info("Running alter sql:\n{}", alterSql);
```
Remove comment.

philipmarzullo64 replied:
Done

erilong commented on symmetric-db/src/main/java/org/jumpmind/db/platform/AbstractDatabasePlatform.java:
```diff
@@ -264,15 +274,33 @@ public void alterTables(boolean continueOnError, IAlterDatabaseInterceptor[] int
-            new SqlScript(alterSql, getSqlTemplate(), !continueOnError, false, false, delimiter, null)
+            Map<String, String> replacementTokens = new HashMap<String, String>();
+            replacementTokens.put(ddlBuilder.getTriggerDelimiterReplacementCharacters(), getDdlBuilder().getDatabaseInfo().getSqlCommandDelimiter());
+            log.info("Running alter sql:\n{}", getAlterSql(alterSql, createTableIncludeApplicationTriggers, replacementTokens, delimiter));
```
Why not just log alterSql?  What does getAlterSql() method do that's any different than that?

philipmarzullo64 replied:
The SqlScript class needs to delimit each SQL statement, and uses semicolon. Also the trigger DDL language has embedded semicolons. So the embedded semicolons are replaced with "&&&&" to allow the semicolon delimiter to work with the SqlScript class. So, in order for the log to show correctly what will actually be executed (since the SqlScript class has "&&&&" in some of the DDL), it has to run through the method to replace "&&&&" with ";".

erilong commented on symmetric-db/src/main/java/org/jumpmind/db/platform/AbstractDatabasePlatform.java:
```diff
@@ -231,22 +235,22 @@ public void createDatabase(Database targetDatabase, boolean dropTablesFirst, boo
     @Override
-    public void alterDatabase(Database desiredDatabase, boolean continueOnError, IAlterDatabaseInterceptor[] interceptors) {
-        alterTables(continueOnError, interceptors, desiredDatabase.getTables());
+    public void alterDatabase(Database desiredDatabase, String symTablePrefix, boolean continueOnError, IAlterDatabaseInterceptor[] interceptors) {
```
We've kept Platform agnostic to anything about Symmetric.  The Dialect was a place to put awareness about Symmetric.  If we just call this variable tablePrefix, is it usable by others?

erilong commented on symmetric-db/src/main/java/org/jumpmind/db/platform/AbstractDatabasePlatform.java:
```diff
@@ -1259,10 +1287,10 @@ protected PermissionResult getAlterSymTablePermission(Database database) {
             database.removeAllTablesExcept();
             database.addTable(alterTable);
-            alterDatabase(database, false);
+            alterDatabase(database, "sym", false);
```
Should "sym" be a tablePrefix variable that is passed in?

philipmarzullo64 replied:
This is just for the permissions available to the database user, and the test table is hardcoded (prefix starts with SYM_)

erilong commented on symmetric-io/src/main/java/org/jumpmind/symmetric/io/data/writer/DefaultDatabaseWriter.java:
```diff
@@ -690,8 +690,18 @@ protected boolean create(CsvData data, boolean withoutDefaults) {
                     }
                 }
             }
+            // if (writerSettings.isCreateTableIncludeApplicationTriggers()) {
```
Remove commented code.

philipmarzullo64 replied:
Done

erilong commented on symmetric-jdbc/src/main/java/org/jumpmind/db/platform/postgresql/PostgreSqlDdlReader.java:
```diff
@@ -408,4 +417,39 @@ public Trigger mapRow(Row row) {
+    
+    @Override
+    public PlatformTrigger getPlatformTrigger(IDatabasePlatform platform, Trigger trigger) {
+        // create or replace function $(schemaName)f$(triggerName)() returns trigger as $function$
```
Remove commented-out code.

philipmarzullo64 replied:
Done

erilong commented on symmetric-db/src/main/java/org/jumpmind/db/platform/AbstractDdlBuilder.java:
```diff
@@ -129,6 +133,7 @@ public abstract class AbstractDdlBuilder implements IDdlBuilder {
     protected boolean sqlCommentsOn = false;
     protected boolean scriptModeOn = false;
     protected String databaseName;
+    private String triggerDelimiterReplacementCharacters = "&&&&";
```
What is the purpose of this?

philipmarzullo64 replied:
The SqlScript class needs to delimit each SQL statement, and uses semicolon. Also the trigger DDL language has embedded semicolons. So the embedded semicolons are replaced with "&&&&" to allow the semicolon delimiter to work with the SqlScript class.

erilong commented on symmetric-db/src/main/java/org/jumpmind/db/platform/AbstractDdlBuilder.java:
```diff
@@ -855,6 +874,53 @@ protected void processChange(Database currentModel, Database desiredModel, CopyC
+            StringBuilder ddl) {
+        ddl.append("DROP FUNCTION ");
+        if (change.getFunction().getCatalogName() != null && change.getFunction().getCatalogName().length() > 0) {
+            ddl.append(change.getFunction().getCatalogName()).append(".");
```
Should this use databaseInfo.getCatalogSeparator()?

philipmarzullo64 replied:
Done

erilong commented on symmetric-db/src/main/java/org/jumpmind/db/alter/ModelComparator.java:
```diff
@@ -326,6 +333,90 @@ public void detectPrimaryKeyChanges(Database sourceModel, Table sourceTable,
+                    PlatformTrigger targetPlatformTrigger = targetTrigger.findPlatformTrigger(platformName);
+                    PlatformTrigger sourcePlatformTrigger = sourceTrigger.findPlatformTrigger(platformName);
+                    boolean triggerTextChanged = ! StringUtils.equals(targetPlatformTrigger.getTriggerText(), sourcePlatformTrigger.getTriggerText());
+                    String targetFunctionText = getFunctionText(targetPlatformTrigger, platformName);;
```
Remove extract semi-colons.

philipmarzullo64 replied:
Done

erilong commented on symmetric-db/src/main/java/org/jumpmind/db/alter/ModelComparator.java:
```diff
@@ -326,6 +333,90 @@ public void detectPrimaryKeyChanges(Database sourceModel, Table sourceTable,
+                }
+            } else {
+                changes.add(new RemoveTriggerChange(sourceTable, sourceTrigger));
+                String platformName = getPlatformName(sourceTrigger);
```
Why does it get the first platform?  Shouldn't it be using the current platform and asking if any matching platform triggers exist for it?

philipmarzullo64 replied:
Put database name in database model objects so ModelComparator can find the explicit platform specific names

Commit:
6690: Code formatting

Commit:
6690: Code formatting

Commit:
6690: Code formatting

Commit:
6690: Code formatting

Commit:
6690: Code formatting

Commit:
feature/6690_send_trigger_ddl: Remove commented out code

Commit:
feature/6690_send_trigger_ddl: remove commented code

Commit:
feature/6690_send_trigger_ddl: use databaseInfo.getCatalogSeparator()
and databaseInfo.getSchemaSeparator() instead of hard-coded ".".

Commit:
feature/6690_send_trigger_ddl: remove extra semicolons

Commit:
feature/6690_send_trigger_ddl: Assign database name to the database
model objects for use by ModelComparator when looking for platform
specific triggers and functions

Commit:
6690_send_trigger_ddl: change symTablePrefix to triggerPrefix (to make
SymmetricDS agnostic) and fix bug (use triggerPrefix, not tablePrefix)

philipmarzullo64 committed:
Merge branch '3.15' into feature/6690_send_trigger_ddl
